### PR TITLE
feat(agents): implement GitHub custom tool handlers (VID-611)

### PR DESCRIPTION
**Pull request by vidbina**
## Summary

- Adds `agents/src/tools/github.ts` with four MCP-compatible tool handlers: `github_get_repo_info`, `github_get_default_branch`, `github_push_branch`, `github_create_pr`
- Follows the same `createXxxToolHandlers(token)` + `xxxToolSchemas` export shape as `linear.ts`
- Replaces direct `GITHUB_TOKEN` env injection in the session container — custom tool dispatch + `limited` networking eliminates the prompt-injection exfiltration surface

## Test plan

- [x] `devenv shell -- pnpm test:run` — all 26 tests pass (9 new in `github.test.ts`)
- [x] Auth header, each endpoint URL/method/body, HTTP 4xx error handling all covered
- [x] `github_push_branch` 4-step flow (ref resolution → tree → commit → ref create) asserted call-by-call

Stacked on: `vidbina/vid-607-try-ant`
Linear: https://linear.app/asabina/issue/VID-611

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: agents/src/tools/github.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        content: f.content,
+      }));
+
+      const treeRes = (await post(`/repos/${owner}/${repo}/git/trees`, {
+        base_tree: baseSha,
+        tree,
+      })) as { sha: string };
+
+      // 3. Create commit
+      const commitRes = (await post(`/repos/${owner}/${repo}/git/commits`, {
```
Comment (vidbina) [inline agents/src/tools/github.ts:74]: missing signiature?

Comment (vidbina) [reply agents/src/tools/github.ts:74]: what is best practice: allocate dedicated sig for machine or reuse human-used sig?

File: agents/tests/tools/github.test.ts
Comment (vidbina) [inline agents/tests/tools/github.test.ts:1]: from gh copilot:

I'll load the PR understanding ability and review the tests in this pull request to evaluate their quality and coverage.Now let me load the pr-reviewer ability since the user is asking for a critical evaluation of the tests:Now let me fetch the PR details and test files to conduct a thorough review:Now let me get the actual file contents to do a more detailed review:## Critical Test Analysis: Solid Foundation but Missing Real-World Scenarios

These tests are **fairly solid** for covering the mechanics of the tool, but they're **surface-level and don't exercise the core logic deeply**. Here's my roast:

### ✅ **What's Good**

1. **The `github_push_branch` orchestration test** (lines 112-167) is the gem here—it validates the 4-step API choreography end-to-end with proper state threading (base SHA → tree SHA → commit SHA → ref creation). This is non-trivial.
2. **Request formation is validated** (headers, URLs, HTTP methods, JSON bodies).
3. **Two error paths** are covered (404 and 401).
4. **All four handlers are exercised** with basic happy paths.

### ❌ **Critical Gaps**

#### **1. No Edge Cases in URL Construction**
The code does **string interpolation on user inputs** without validation:
```typescript
get(`/repos/${owner}/${repo}`)
post(`/repos/${owner}/${repo}/git/refs`, { ref: `refs/heads/${branch}`, ... })
```
- **Missing tests**: What happens if `owner`, `repo`, or `branch` contain special characters (slashes, spaces, etc.)?
- What if `branch` is an invalid ref name (e.g., `refs/heads/` or `../../../etc/passwd`)?
- The tests use only safe strings like `"acme"` and `"widget"`—no adversarial input.

#### **2. `github_push_branch` Logic Holes Not Tested**
```typescript
const fileList = files as Array<{ path: string; content: string }>;
const tree = fileList.map((f) => ({
  path: f.path,
  mode: '100644',
  type: 'blob',
  content: f.content,
}));
```
- **Missing**: Empty files array—does it handle the edge case? Test only uses 1 file.
- **Missing**: Files with problematic paths (absolute paths like `/etc/passwd`, `.git/config`, deeply nested paths, special characters).
- **Missing**: Large content that might exceed GitHub API limits (the code blindly sends whatever arrives).
- **Missing**: Duplicate file paths—will GitHub reject? Does the code handle it gracefully?

#### **3. Response Parsing is Fragile**
```typescript
const info = (await get(`/repos/${owner}/${repo}`)) as { default_branch: string };
return { default_branch: info.default_branch };
```
- **Missing**: What if the response doesn't have `default_branch`? The code casts but doesn't validate.
- **Missing**: What if `object.sha` is missing in the ref lookup? Uncaught error.
- No tests for partial/malformed API responses.

#### **4. No Network Failure Scenarios**
- Tests mock fetch as always resolving—no timeout simulation.
- No handling of rate limiting (HTTP 429), redirects (3xx), or server errors (5xx).
- The error handling test assumes `res.text()` succeeds—what if that rejects?

#### **5. Zero Validation of User Inputs**
- No test for `null`/`undefined` fields in handler arguments.
- No test for type mismatches (e.g., `files` as a string instead of array).
- The `ToolInput` type is `Record<string, unknown>` but handlers assume correct shape—**type unsafety is real**.

#### **6. `github_create_pr` Skips Validation**
```typescript
github_create_pr: ({ owner, repo, title, body, head, base }) =>
  post(`/repos/${owner}/${repo}/pulls`, { title, body, head, base }),
```
- **Missing**: What if `title` is empty or `body` is null?
- **Missing**: What if `head` and `base` are the same branch?
- Test passes any values without checking API realism.

#### **7. Commit Message is Hardcoded**
```typescript
message: `chore: push branch ${branch}`,
```
- **Missing**: What if `branch` injection breaks the message format? (Not tested)
- **Missing**: Should this be customizable? Test assumes current behavior is correct.

#### **8. No Integration Semantics**
- Tests verify HTTP calls but not GitHub semantics. Example: creating a branch that already exists—GitHub returns 422, but the test doesn't cover this.
- No test for dependent-call failures (e.g., step 1 succeeds but step 2 fails mid-flow—is state left inconsistent?).

### 📋 **What Should Be Tested**

```typescript
describe('Input Validation', () => {
  it('rejects branch names with invalid characters');
  it('rejects empty owner/repo strings');
  it('rejects files with absolute paths');
  it('rejects files with null content');
  it('handles empty files array');
  it('handles files array with duplicates');
});

describe('Response Parsing', () => {
  it('throws if base ref response lacks object.sha');
  it('throws if default_branch is missing from repo response');
  it('throws if tree response lacks sha');
});

describe('Network Errors', () => {
  it('handles fetch timeout gracefully');
  it('handles 429 rate limiting');
  it('handles 5xx server errors with retry?');
});

describe('GitHub API Edge Cases', () => {
  it('handles 422 when branch already exists');
  it('handles creating PR when head/base are same');
  it('handles files exceeding size limits');
});
```

### 🎯 **Verdict**

These tests **pass the smell test** (all handlers are called, basic happy path works, orchestration is validated), but they're **not stress-tested against real-world failure modes**. They'll catch regressions in happy-path logic but won't prevent:
- Security issues from unsanitized input
- Runtime crashes from malformed API responses
- Silent failures from edge-case handling

**Risk Level**: Medium. The code is unlikely to crash in normal use, but edge cases could surprise you in production.

File: agents/src/tools/github.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        message: `chore: push branch ${branch}`,
```
Comment (vidbina) [inline agents/src/tools/github.ts:75]: drop magic string and expose through param such that the caller defines this.

alt: we can parameterize the first line and boilerplate some parts of the message to include the Co-Authored line with just the name of the co-author as another param, but this may become too rigid. Some commits will not need this or not know a valid value for this (since agent email address will be unknown) and in this case, we should not force a dummy value. Final design tbd, let's rethink this carefully.

In the classic case I will expect:

```
SEMANTIC_COMMIT_LINE

COMMIT_SUMMARY

Co-Authored-By: AGENT_BYLINE
```

Where only the first line is a hard must-have. The appearance of the other components of the aforementioned template are case-dependent. Simple package updates may lack summaries or co-author bylines (e.g.: regular package update jobs that never involve any AI-supported reasoning).

**Review by vidbina — COMMENTED** [review]
decent attempt and an extensoin to the try ant scope https://github.com/vidbina/dotfiles/pull/37 but still has topics to be /linearissue ed and tagged as Debt or Enchancement (I bet many of the smell issues are just Debt, but topics like adding sigs is definitely more Enhancement). Note that #37 has already drifted quite out of scope (due to its initial mandate being quite broad) so let's not try to cram too much into existing PR scopes but relegate to future work what makes sense to isolate. Expectation is that we will have to rewrite the whole setup here as we learn what does and doesn't work. :-/
